Make adding text via "add new text" or (T+) to use same default font

ezxyz opened the pull request:
## Summary

Made adding text via "add new text" or (T+) to use same default font

## Relevant Technical Choices
- Default font is the same settings as body text, however doesn't use the same PRESETS constant. (This is so that it isn't tied to presets when we implement the ability to edit presets in the future).
- Created a new textPresets.js to deal with Presets and Default Font Style.
- in textPane.js removed "getPresetById" since it wasn't being used anymore

Addresses #1204 partially (the rest should be split out to a new issue).

Commit:
Made adding text via "add new text" or (T+) to use same default font

- Default font is the same as body text.
- Created a new textPresets.js to deal with Presets and Default Font Style.
- in textPane.js removed "getPresetById" since it wasn't being used anymore

Commit:
Removed unused import.

swissspidy commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,76 @@
+  content: __('Fill in some text', 'web-stories'),
+  fontSize: dataFontEm(1.1),
+  fontWeight: 400,
+  font: {
```
There's already `TEXT_ELEMENT_DEFAULT_FONT` in `assets/src/edit-story/app/font/defaultFonts.js`  which we can use.

ezxyz replied:
Do you mean we should import the TEXT_ELEMENT_DEFAULT_FONT and use it in textPresets.js or to move the DEFAULT_FONT in this file to the defaultFonts.js

ezxyz replied:
TEXT_ELEMENT_DEFAUL_FONT seems to have a series of different weights instead of just one weight? I'm a bit confused about how the TEXT_ELEMENT_DEFAULT_FONT can be used in the same way.

dvoytenko replied:
Not sure why this got resolved. But IMHO this is a critical part to address. Yes, a standard font definition should have all the weights/styles/variants that it supports so that we do not have to query any database when processing it.

dvoytenko replied:
To clarify, we need to import `TEXT_ELEMENT_DEFAULT_FONT` from the `defaultFonts.js` and use it everywhere here.

ezxyz replied:
PTAL. Was the new change what you meant?

swissspidy commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,76 @@
+ */
+import { dataFontEm } from '../../../../units';
+
+const DEFAULT_FONT = {
```
This name is a bit misleading as it's not about the font.

ezxyz replied:
should it be DEFAULT_PRESET ?

ezxyz replied:
I've changed the name to DEFAULT_PRESET.

barklund commented on assets/src/edit-story/components/library/panes/text/textTab.js:
```diff
@@ -82,8 +81,7 @@ function TextTab(props) {
     insertElement('text', {
-      content: __('Fill in some text', 'web-stories'),
-      fontSize: DEFAULT_FONT_SIZE,
+      ...DEFAULT_FONT('default-text'),
```
Is it a function?

ezxyz replied:
Sorry about that, I missed that from when I initially made it a function.

Commit:
Fixed mistake with DEFAULT_FONT.

Commit:
Changed DEFAULT_FONT name to DEFAULT_PRESET

Commit:
Update textTab.js

swissspidy commented on assets/src/edit-story/components/library/panes/text/textTab.js:
```diff
@@ -82,8 +81,7 @@ function TextTab(props) {
-      content: __('Fill in some text', 'web-stories'),
-      fontSize: DEFAULT_FONT_SIZE,
+      ...DEFAULT_PRESET,
       width: DEFAULT_ELEMENT_WIDTH,
```
Does this constant's value need to be updated? According to the issue description / Figma text elements should be 360 wide.

ezxyz replied:
As per discussion on slack we should split this part into a new issue.

swissspidy commented on assets/src/edit-story/components/library/panes/text/textPane.js:
```diff
@@ -107,8 +59,7 @@ function TextPane(props) {
-                ...getPresetById('subheading'),
-                content: __('Fill in some text', 'web-stories'),
+                ...DEFAULT_PRESET,
                 width: DEFAULT_ELEMENT_WIDTH,
```
Does this constant's value need to be updated? According to the issue description / Figma text elements should be 360 wide.

ezxyz replied:
Same as above.

swissspidy commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,76 @@
+ */
+import { dataFontEm } from '../../../../units';
+
+const DEFAULT_PRESET = {
```
The issue mentions  default values for text alignment, box size, line height, etc.

Those seem to be all missing from this PR? Or is this defined elsewhere?

ezxyz replied:
So the box size will be in a different issue as discussed.

However text alignment is actually left aligned in figma (even though the issue mistakenly says center aligned). Since thats the default alignment, I don't think we need to add that.

I added line height in a new commit.

Commit:
Merge branch 'master' into new-text

Commit:
Added line height, fixed the font size so that it is 16px (1em).

swissspidy commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,80 @@
+    family: 'Roboto',
+    service: 'fonts.google.com',
+  },
+  lineHeight: '24px ',
```
Errant space?

ezxyz replied:
Done

swissspidy reviewed (approved):
LGTM at first glance, but please get the final approval from the workspace pod

Commit:
Removing stray space

miina commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,80 @@
+    family: 'Roboto',
+    service: 'fonts.google.com',
+  },
+  lineHeight: '24px',
```
Note that the `lineHeight` used for text element is not in pixels, it's just a number. So this should probably be calculated based on the font size to get `24px`.

dvoytenko replied:
Correct. I believe it's still a fractional value. But please double check. There was a plan to switch it to a percent value.

dvoytenko replied:
Though, why do we need it here at all? Why not just go with `auto`?

ezxyz replied:
This is due to the UX requirement in the bug and in figma of a 24px line height. Since 1em is 16px, then it should be 1.5

dvoytenko replied:
Ok, I see. This is conflated with the https://github.com/google/web-stories-wp/issues/1820 issue, which is fairly broader. I recommend dropping `lineHeight` from here (e.g. assuming the auto value will be addressed at some point) and instead focusing on other nuances.

ezxyz replied:
Removed.

miina commented on assets/src/edit-story/components/library/panes/text/textTab.js:
```diff
@@ -82,8 +81,7 @@ function TextTab(props) {
     insertElement('text', {
-      content: __('Fill in some text', 'web-stories'),
-      fontSize: DEFAULT_FONT_SIZE,
+      ...DEFAULT_PRESET,
```
Would be ideal to have a test for `textTab` which would check that on clicking `QuickAction`, `insertElement` is called with the correct attributes.

ezxyz replied:
Added a new test. PTAL

dvoytenko commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,80 @@
+  id: 'default-text',
+  content: __('Fill in some text', 'web-stories'),
+  fontSize: dataFontEm(1),
+  fontWeight: 400,
```
Font weights have been dropped entirely, and moved to the content.

ezxyz replied:
Fixed.

dvoytenko reviewed (changes requested):
Generally all looks good. Just a few comments.

Commit:
Update textPresets.js

swissspidy commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,77 @@
+ * Internal dependencies
+ */
+import { dataFontEm } from '../../../../units';
+import { TEXT_ELEMENT_DEFAULT_FONT } from '../==../../app/font/defaultFonts';
```
That's some odd import 🙂

ezxyz replied:
D'oh 🤦 (Fixed)

Commit:
Fixed path to defaultFonts

Commit:
Removed line height.

Commit:
Adding tests for textTab and textPane for adding default text

Commit:
Merge branch 'master' into new-text

miina commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,76 @@
+    'Fill in some text',
+    'web-stories'
+  )}</span>`,
+  fontSize: dataFontEm(1),
```
Where does this value come from btw?
For example, on my 13'' screen, using `dataFontEm(1)` will display tiny, probably the real size in pixels will be less than 10px. That might not be even accessible anymore due to the small font size depending on the screen.

Assuming the Story is viewed in a mobile device it's likely that this will make the default quite small in the FE as well.

miina replied:
FI: Tested it out on my 13'' screen:
- The real font size in the editor is `7.63636px`
- In the FE checking with the Chrome device toolbar for iPhone X, the real font size is around `12px`.

ezxyz replied:
1) In regards to using 1 em: 
I found some sources saying that the defualt pixel size is usually 16px (e.g. https://www.w3schools.com/tags/ref_pxtoemconversion.asp). So I assumed because the specs said 16px in the issue, that we were supposed to put in 1em so that it scales according to device. (16px on a screen with high pixel density would be extremely small)

2) In regards to em in general:
Shouldn't EM scale based on how small your screen is? So in a small screen the em should rescale to a value based on your screen size and pixel density?

Regardless the other body text in that file are currently using 1.1 em and it already exists in the current codebase. You can have a look at the current story editor and see what the body font preset looks like on your 13 in screen to test this out. If it actually does look very small, then it might be worth rethinking all of the fonts sizes in general that use em?

ezxyz replied:
Also I realised I didn't see your second comment before I commented the above. I think if the real font editor on the current build seems small for you this is an issue we should bring to the team and see how to best approach it.

miina replied:
Yep, saw the `1.1em` for the body text and that, too, seemed to be quite tiny in the editor on a smaller screen. Perhaps the idea for the "body text" is to fit a lot of text on a Page which might be fine in the FE then.

For the default Text element, this value definitely seems too small though and not really accessible, it's quite eye-straining on my 13'', and also looks too small when testing FE with mobile.

Perhaps it could make sense to just test which font size would look good for a default Text element on the FE, e.g. for some popular mobile device and set the default value based on that?

Thoughts?

dvoytenko replied:
`1em` is currently about `5%` of the page's height, or should be `14px` in data pixels. Depending on the actual screen size is could be any value, but it's always about `5%` of the page's height.

ezxyz replied:
If 1em should be about 14px in pixels, there might be a larger issue with the em sizes since @miina saw it at 7.6px. It's possible the current DEFAULT_EM calculation is not working very well in smaller screen sizes. I think we should open a new issue about this as it affects all fonts with font size in em.

ezxyz replied:
Considering that 1em is 14px in pixels, should I just put in 1.1em for this PR, and we can open a new issue about fixing font sizes in general in smaller screens to investigate this further? I'd prefer the default text font to be relative to the rest of the fonts and then fix the em value, rather than just have this one be really big compared to the other ones while we wait to fix the rest.

I think this is beyond the scope of this if the em value itself is too small. Thoughts?

ezxyz replied:
I added the font size of 1.5 em as per conversation on slack with @miina PTAL.

samitron7 replied:
@miina In your initial comment, what value are you looking at to get to the 10px value? @o-fernandez responsiveness is an issue that requires spec-ing out and prioritize.

miina replied:
@samitron7 In the editor, for getting the `7.63636px` font size, the used value for the default font was `1em` (14 as the font size in Design Panel). In case of larger screens, the font in px is larger as well.

Since we're not using actual pixels in the editor then it's not that straightforward to set the default value to `16px`. It might be good to specify the device (or viewport size) where the font size in pixels should be `16` in the front-end -- since on different devices the font size will differ as well. Thoughts?

dvoytenko replied:
IMHO we should keep this at `1` for now. `1em` for us simply means the "default size". Because the story is proportionally responsive, the meaning of `1em` should not change based on how much the editor is zoomed in/out, or how much the output AMP story is zoomed in/out. Currently, `1em` is set to be about `5%` of the story page, i.e. that maps to ~14 lines of text that would fit on the page. If we like some other value that would work better as a default, let's pick one and update it for the whole editor. We'll just need to change a single [`DEFAULT_EM` constant](https://github.com/google/web-stories-wp/blob/b93f5b8d33e17eec5e76c4cc975951f90a809313/assets/src/edit-story/constants.js#L46). As such, I see this problem mostly separate from the original bug report. And, the value of `DEFAULT_EM` should definitely be decided by @samitron7.

ezxyz replied:
I agree with @dvoytenko that it should be 1em for now since then everything is proportional. I can change it back to 1em if @miina is happy with it. LMK what you think @miina

miina replied:
I agree that from our side it makes sense to have it `1em`, however, from the product's perspective it does not make sense to have a default value which is not usable on smaller screens. As long as it gets fixed soon, it's OK by me but let's follow up on this with @samitron7 to ensure it's not forgotten.

ezxyz replied:
Added the 1em fix. PTAL.

Commit:
Changed text fontsize as per offline conversation with mina

Commit:
minor changes to the test descriptions

miina commented on assets/src/edit-story/components/library/test/text/textTab.js:
```diff
@@ -0,0 +1,71 @@
+    expect(insertElement).toHaveBeenCalledTimes(1);
+    expect(insertElement).toHaveBeenCalledWith('text', {
+      ...DEFAULT_PRESET,
+      width: 160,
```
Q: Shouldn't the width here be the same as when added via QuickAction or the shortcut link?

ezxyz replied:
I decided to leave the width here as the same because in the issue we were going to move changing the textbox width and position out of the scope of this issue.

Would you like me to change the width to be consistent across both of them? Which one should I use? 

I'm not sure what width to use (the issue says 360px but that seems wrong considering in the Figma specs [text > write specs > recommended text width] the textbox width is capped at 332px. Both half the width of the page and 160px seems fine to me. @samitron7  thoughts?

miina replied:
It just seems that it would make sense to have these consistent, that's all, choosing any value should be OK if this will be addressed separately later anyway to finalize. Thoughts?

ezxyz replied:
Sounds good. I chose the 140 px value from TextTab since it better spans the default text. I then moved the default element width value to the presets and added them directly there instead. PTAL.

Commit:
Moved all the default element widths to the presets

Also set a default width for DEFAULT_PRESET that is different, that approximately spans the text inside it (based on the existing default width from textTab.js)

miina reviewed (approved):
LGTM, the text added via the + icon and the `QuickAction` produce the same results now.

dvoytenko commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,84 @@
+      'web-stories'
+    )}</span>`,
+    fontSize: dataFontEm(2),
+    font: {
```
This is not nearly enough data, unfortunately, to properly display/render the text element. See `TEXT_ELEMENT_DEFAULT_FONT` for what that usually requires. Can't we use `TEXT_ELEMENT_DEFAULT_FONT` here as well?

dvoytenko replied:
Please use `TEXT_ELEMENT_DEFAULT_FONT` in place of "Open Sans" everywhere in this file.

ezxyz replied:
Done.

dvoytenko commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,84 @@
+      'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
+      'web-stories'
+    ),
+    fontSize: dataFontEm(1.1),
```
Why `1.1`?

ezxyz replied:
I didn't change this, this was the dataFontEm from before. So I'm not really familiar with the decision around this. I thought it was out of scope to change the rest of the fonts.

dvoytenko commented on assets/src/edit-story/components/library/panes/text/textPresets.js:
```diff
@@ -0,0 +1,84 @@
+// By default, the element should be 50% of the page.
+const DEFAULT_ELEMENT_WIDTH = PAGE_WIDTH / 2;
+
+const DEFAULT_PRESET = {
```
Why is `DEFAULT_PRESET` at all different from the "normal body" preset?

dvoytenko reviewed (approved):
LGTM with one ask:

> Please use TEXT_ELEMENT_DEFAULT_FONT in place of "Open Sans" everywhere in this file.

Commit:
Using text element default font for every font.
